### DISCUSSION TITLE
yaml module not found

### DISCUSSION THREAD

**Discussion by WulfHans** [idaholab/moose#27133] [Q&A Getting Started] [answered]
Install on a clean machine (linux, conda environment) failed with 
`ModuleNotFoundError: No module named 'yaml'`
The error already occured during build, but the build still finished. However, running any tests fails instantly.

The common advice on the net is to install pyyaml. However, this is already installed:
```
(moose) wuha@pontos:/home/.local/wuha$ conda list pyyaml
# packages in environment at /home/.local/wuha/miniforge/envs/moose:
#
# Name                    Version                   Build  Channel
pyyaml                    6.0.1           py310h2372a71_1    conda-forge
```

The python seems to come from the correct location (miniforge), but yaml is indeed missing:
```
(moose) wuha@pontos:/home/.local/wuha$ which python
/home/wuha/miniforge/bin/python
(moose) wuha@pontos:/home/.local/wuha$ python
Python 3.10.13 | packaged by conda-forge | (main, Oct 26 2023, 18:07:37) [GCC 12.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import yaml
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ModuleNotFoundError: No module named 'yaml'
```
Typical attempts like updating everything or just redoing several steps did not solve it either. 
Any recommendation? Thanks in advance!

**Comment by GiudGiud** [comment]
Hello

you can also install yaml.
`conda install yaml`

Guillaume

**Reply by milljm** [reply]
I think its actually `pyaml` you want. This will get the shared libraries for `yaml` and the Python module to interface with it `import yaml`. Can you believe all these exist and are different!?:

```pre
conda install yaml pyaml pyyaml
```
:)

**Reply by WulfHans** [reply]
yaml and pyyaml were installed, pyaml was missing. Issues is solved. Please consider adding the module to the dependency package you get with
`conda create -n moose moose-dev=2024.03.15`

**Reply by milljm** [reply]
Well interesting. Something else must have been wrong because that's already a part of `moose-dev`. "moose-dev" depends on "moose-tools", which brings in the proper `yaml` modules. I am not sure how you ended up not having it.

The results of my conda list after ` conda create -n moose moose-dev=2024.03.15`:
```bash
$ conda list | grep yaml
pyaml                     23.12.0            pyhd8ed1ab_0    conda-forge
pyyaml                    6.0.1           py310h2372a71_1    conda-forge
ruamel.yaml               0.18.6          py310h2372a71_0    conda-forge
ruamel.yaml.clib          0.2.8           py310h2372a71_0    conda-forge
yaml                      0.2.5                h7f98852_2    conda-forge
```

**Comment by WulfHans** [comment]
Sorry, I have to reopen the issue. 
Today, it does not work again. I have no idea, what I changed.

It seems as if all files and packages are present, but it does not work nevertheless:

```
(moose) wuha@pontos:/home/.local/wuha/PRIVAT/moose-projects/moose$ conda list yaml
# packages in environment at /home/.local/wuha/miniforge/envs/moose:
#
# Name                    Version                   Build  Channel
pyaml                     23.12.0            pyhd8ed1ab_0    conda-forge
pyyaml                    6.0.1           py310h2372a71_1    conda-forge
ruamel.yaml               0.18.6          py310h2372a71_0    conda-forge
ruamel.yaml.clib          0.2.8           py310h2372a71_0    conda-forge
yaml                      0.2.5                h7f98852_2    conda-forge
(moose) wuha@pontos:/home/.local/wuha/PRIVAT/moose-projects/moose$ python
Python 3.10.13 | packaged by conda-forge | (main, Oct 26 2023, 18:07:37) [GCC 12.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import yaml
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ModuleNotFoundError: No module named 'yaml'
```

**Reply by GiudGiud** [reply]
what does `which python` return?

**Reply by milljm** [reply]
This feels like a PATH issue somehow.
What does the following report?

```bash
which python python3 conda
```

**Reply by WulfHans** [reply]
```
(moose) wuha@pontos:~/PRIVAT/moose-projects/moose_pattern$ which python
/home/wuha/miniforge/bin/python
(moose) wuha@pontos:~/PRIVAT/moose-projects/moose_pattern$ which python3
/home/wuha/miniforge/bin/python3
(moose) wuha@pontos:~/PRIVAT/moose-projects/moose_pattern$ which conda
/home/wuha/miniforge/bin/conda
```

No big surprises

**Reply by milljm** [reply]
It is interesting that `conda` is being reported as a non-function. It's possible Conda is not being initialized properly? Can you print out the conda info page, and then for troubleshooting measures, I want to re-initialize Conda:


```bash
conda info --all
conda deactivate
source activate /home/wuha/miniforge                     # this might fail
source activate /home/.local/wuha/miniforge/envs/moose   # this might fail
python -c 'import yaml; print(yaml.__file__)'
```

**Reply by WulfHans** [reply]
That gives quite a wall of text...
```

(moose) wuha@pontos:/home/.local/wuha/PRIVAT/moose-projects/moose$ conda info --all

     active environment : moose
    active env location : /home/.local/wuha/miniforge/envs/moose
            shell level : 3
       user config file : /home/wuha/.condarc
 populated config files : /home/.local/wuha/miniforge/.condarc
                          /home/wuha/.condarc
          conda version : 24.1.2
    conda-build version : not installed
         python version : 3.10.13.final.0
                 solver : libmamba (default)
       virtual packages : __archspec=1=haswell
                          __conda=24.1.2=0
                          __glibc=2.36=0
                          __linux=6.1.0=0
                          __unix=0=0
       base environment : /home/.local/wuha/miniforge  (writable)
      conda av data dir : /home/.local/wuha/miniforge/etc/conda
  conda av metadata url : None
           channel URLs : https://conda.software.inl.gov/public/linux-64
                          https://conda.software.inl.gov/public/noarch
                          https://conda.anaconda.org/conda-forge/linux-64
                          https://conda.anaconda.org/conda-forge/noarch
          package cache : /home/.local/wuha/miniforge/pkgs
                          /home/wuha/.conda/pkgs
       envs directories : /home/.local/wuha/miniforge/envs
                          /home/wuha/.conda/envs
               platform : linux-64
             user-agent : conda/24.1.2 requests/2.31.0 CPython/3.10.13 Linux/6.1.0-17-amd64 debian/12.5 glibc/2.36 solver/libmamba conda-libmamba-solver/24.1.0 libmambapy/1.5.7
                UID:GID : 54477:4000
             netrc file : None
           offline mode : False


# conda environments:
#
base                     /home/.local/wuha/miniforge
moose                 *  /home/.local/wuha/miniforge/envs/moose

sys.version: 3.10.13 | packaged by conda-forge | (mai...
sys.prefix: /home/.local/wuha/miniforge
sys.executable: /home/.local/wuha/miniforge/bin/python
conda location: /home/.local/wuha/miniforge/lib/python3.10/site-packages/conda
conda-build: None
conda-env: /home/.local/wuha/miniforge/bin/conda-env
user site dirs: ~/.local/lib/python2.7

BOOST_BUILD_PATH: /home/wuha/PRIVAT/OptiSlang/develop/boost/boost_1_72_0/tools/build
CIO_TEST: <not set>
CMAKE_PREFIX_PATH: /home/.local/wuha/miniforge/envs/moose:/home/.local/wuha/miniforge/envs/moose/x86_64-conda-linux-gnu/sysroot/usr
CONDA_BUILD_SYSROOT: /home/.local/wuha/miniforge/envs/moose/x86_64-conda-linux-gnu/sysroot
CONDA_DEFAULT_ENV: moose
CONDA_EXE: /home/.local/wuha/miniforge/bin/conda
CONDA_PREFIX: /home/.local/wuha/miniforge/envs/moose
CONDA_PREFIX_1: /home/.local/wuha/mambaforge3
CONDA_PREFIX_2: /home/.local/wuha/miniforge
CONDA_PROMPT_MODIFIER: (moose) 
CONDA_PYTHON_EXE: /home/.local/wuha/miniforge/bin/python
CONDA_ROOT: /home/.local/wuha/miniforge
CONDA_SHLVL: 3
CONDA_TOOLCHAIN_BUILD: x86_64-conda-linux-gnu
CONDA_TOOLCHAIN_HOST: x86_64-conda-linux-gnu
CURL_CA_BUNDLE: <not set>
C_INCLUDE_PATH: /home/.local/wuha/miniforge/envs/moose/include
GAPS_PATH: /home/wuha/PRIVAT/Identifikation/ami/gaps/bin/
LD_LIBRARY_PATH: /usr/local/cuda/lib64/:/usr/local/cuda-7.0/lib64:/lib:/usr/local/mvapich2/lib/:/usr/local/cuda/lib64/:/usr/local/cuda-7.0/lib64:/lib:/usr/local/mvapich2/lib/::/usr/local/cuda-6.5/lib64:/usr/local/cuda-6.5/lib64
LD_PRELOAD: <not set>
PATH: /home/wuha/miniforge/bin:/usr/local/cuda-7.0/bin:MPI_HOME_DIR/bin:/home/.local/wuha/miniforge/envs/moose/bin:/home/.local/wuha/mambaforge3/condabin:/home/wuha/miniconda3/bin:/usr/local/cuda-7.0/bin:MPI_HOME_DIR/bin:/usr/local/bin:/usr/bin:/bin:/usr/local/games:/usr/games:/home/wuha:/home/wuha/bin:/usr/local/lib:/usr/local/bin:/sbin:/opt/software/MARC/marc2007r1:/opt/software/ANSYS/v121/ansys/bin:/usr/local/cuda-6.5/bin:/home/wuha/PRIVAT/OptiSlang/develop/boost/boost_1_72_0:/home/wuha/PRIVAT/antraege/DFG_KogniProd/Vorarbeiten/fibergen/bin:/home/wuha:/home/wuha/bin:/usr/local/lib:/usr/local/bin:/sbin:/opt/software/MARC/marc2007r1:/opt/software/ANSYS/v121/ansys/bin:/usr/local/cuda-6.5/bin:/home/wuha/PRIVAT/OptiSlang/develop/boost/boost_1_72_0:/home/wuha/PRIVAT/antraege/DFG_KogniProd/Vorarbeiten/fibergen/bin:/home/.local/wuha/miniforge/envs/moose/wasp/bin:/home/.local/wuha/miniforge/envs/moose/wasp/bin:/home/.local/wuha/miniforge/envs/moose/wasp/bin
QTWEBENGINE_DICTIONARIES_PATH: /usr/share/hunspell-bdic/
REQUESTS_CA_BUNDLE: <not set>
SSL_CERT_FILE: <not set>

```

Lets continue:
```
(moose) wuha@pontos:/home/.local/wuha/PRIVAT/moose-projects/moose$ conda deactivate
(base) wuha@pontos:/home/.local/wuha/PRIVAT/moose-projects/moose$ source activate /home/wuha/miniforge
(/home/wuha/miniforge) wuha@pontos:/home/.local/wuha/PRIVAT/moose-projects/moose$ source activate /home/.local/wuha/miniforge/envs/moose
(moose) wuha@pontos:/home/.local/wuha/PRIVAT/moose-projects/moose$ python -c 'import yaml; print(yaml.__file__)'
/home/.local/wuha/miniforge/envs/moose/lib/python3.10/site-packages/yaml/__init__.py
```

And now things work:
```
Python 3.10.8 | packaged by conda-forge | (main, Nov 22 2022, 08:23:14) [GCC 10.4.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import yaml
>>> 
```

I have no idea, what is different now.

**Reply by milljm** [reply]
Also, are you interfacing with your terminal via any IDE? Like VS Code? Just curious! I know my integrated VS Code terminal exhibits very strange behavior sometimes with Conda.

**Reply by WulfHans** [reply]
Nope, just pure bash

**Reply by milljm** [reply]
> Nopw, just pure bash

Okay!

Well some of this is interesting:
```pre
CONDA_PREFIX_1: /home/.local/wuha/mambaforge3
CONDA_PREFIX_2: /home/.local/wuha/miniforge
```
It might be that there are two instances of Conda possibly fighting eachother. I don't know what the right solution here is, except to remove one. Probably the `mambaforge` variant, as that is a deprecated Conda instance as of last September. You probably have that due to MOOSE using mambaforge last year :)

**Reply by WulfHans** [reply]
There was a version with mambaforge on the previous computer, with a different system etc. 
I copied lots of data to the new one, but I am not aware of copying anything related to conda.
There is no reference to mambaforge in bashrc. This directory (`/home/.local/wuha/mambaforge3`) does not exist. I have really no idea what to remove.

**Reply by milljm** [reply]
Can you print the following two files?

```pre
cat ~/.bashrc
cat ~/.bash_profile
```

It seems mambaforge3 is going to be somewhere... since I also notice `conda info --all` printed out your PATH variable, and it too is listing a mambaforge3/bin PATH. Of course, missing directories among PATH should be okay, and therefor if this location physically does not exist, then I doubt mambaforge3 is getting in the way.

Still, there is some oddness. A proper installation of Conda, should result in the following `which` returns:

```bash
❯ which python python3 conda
/Users/milljm/miniforge3/envs/moose/bin/python
/Users/milljm/miniforge3/envs/moose/bin/python3
conda () {
	\local cmd="${1-__missing__}"
	case "$cmd" in
		(activate | deactivate) __conda_activate "$@" ;;
		(install | update | upgrade | remove | uninstall) __conda_exe "$@" || \return
			__conda_reactivate ;;
		(*) __conda_exe "$@" ;;
	esac
}
```
where `conda` is a function, and not a path to the actual binary which is being reported by your `which conda` statement.

The reason why its working "now" is because we sourced the moose environment manually, properly. But we shouldn't have to do this. `conda activate moose` should suffice.

**Reply by milljm** [reply]
Oh, I missed a Conda:

```pre
PATH: /home/wuha/miniforge/bin:/usr/local/cuda-7.0/bin \
MPI_HOME_DIR/bin:/home/.local/wuha/miniforge/envs/moose/bin: \
/home/.local/wuha/mambaforge3/condabin:/home/wuha/miniconda3/bin:
```
Looks like you might have 3 (past) instances of Conda. We should make sure this `/home/wuha/miniconda3` location is also not present. And if it is, figure out how to consolidate what you may need from this Conda instance, into Miniforge3.

**Reply by WulfHans** [reply]
None of these directories does exist. But it seems I have research methods to get rid of incredibly old entries in my PATH. I have no idea where these come from - there is nothing in bashrc.

On Topic: I rather dislike posting my complete bashrc here. If it is required at the end, I will reconsider it.

But I can easily post the part related to conda:

```
# >>> conda initialize >>>
# !! Contents within this block are managed by 'conda init' !!
__conda_setup="$('/home/.local/wuha/miniforge/bin/conda' 'shell.bash' 'hook' 2> /dev/null)"
if [ $? -eq 0 ]; then
    eval "$__conda_setup"
else
    if [ -f "/home/.local/wuha/miniforge/etc/profile.d/conda.sh" ]; then
        . "/home/.local/wuha/miniforge/etc/profile.d/conda.sh"
    else
        export PATH="/home/.local/wuha/miniforge/bin:$PATH"
    fi
fi
unset __conda_setup
# <<< conda initialize <<<
```

I also have this line:
`export PATH=$HOME/miniforge/bin:$PATH`
Is this a bad idea?

**Reply by WulfHans** [reply]
I realized that the successful run was remote, while the issues were on-size. At the end, I just decided to just reboot (what rarely happens otherwise). Now the path looks like

`/home/.local/wuha/miniforge/bin:/home/.local/wuha/miniforge/condabin:MPI_HOME_DIR/bin:/usr/local/bin:/usr/bin:/bin:/usr/local/games:/usr/games:/home/wuha:/home/wuha/bin:/usr/local/lib:/usr/local/bin:/sbin:/home/wuha/PRIVAT/OptiSlang/develop/boost/boost_1_72_0:/home/wuha/PRIVAT/antraege/DFG_KogniProd/Vorarbeiten/fibergen/bin`

And also the stray references to mambaforge in `conda info --all` are gone.
Seems like my terminal stores some persitent info somewhere else that bashrc (?) . 

All seems to work now. I will test again on-site tomorrow and mark it as resolved, if everything works. 

Btw: conda is still not a function:
```
(moose) wuha@pontos:~/PRIVAT/moose-projects/moose_pattern$ which python conda
/home/.local/wuha/miniforge/envs/moose/bin/python
/home/.local/wuha/miniforge/condabin/conda
```

**Reply by milljm** [reply, marked as answer]
Conda not being a function is probably fine... And I think its coming from you having that in your PATH permanently. e.g the `export PATH=$HOME/miniforge/bin:$PATH` line.

The persistence can happen if you are connecting to your remote machine via `screen`, or `tmux`. Where the originating bash shell never dies. Useful! But in this case, it needed to be re-spawned! :)

**Reply by milljm** [reply]
I am busy trying to get a diagnostic script available for issues like your (https://github.com/idaholab/moose/pull/26901). I am curious what this might have reported had it been available. Only posting in case you'd like to try my branch (if you're comfortable doing that, and if the problem persists).

**Reply by WulfHans** [reply]
I would like to help you with your diagnostic script, but the problem is solved and I really dont want to mess it up again. The culprit probably was `tmux` - I switch between home-office and on-site  a lot and use it extensively. I will set up moose one other computers as well in the future - if I encounter some issues again, I will give it a try.
For now, I mark it as solved again. Thanks a lot for your quick and persistent support!
